### PR TITLE
[KBFS-1654] Fix lots of things to make v3 testing easier

### PR DESCRIPTION
**Pull request by akalin-keybase**
Add a defaultClientMetadataVer constant to control
the V3 switchover. Use it everywhere.

Add MakeInitialBareRootMetadataV{2,3} functions
and MakeInitialBareRootMetadata function that
takes a MetadataVer. Use it everywhere.

Reduce the MutableBareRootMetadata interface a bit.

Rename signMD to SignBareRootMetadata and use
it everywhere. Also add test coverage for the V3 signing
and implement it. Make MakeFinalCopy take the finalizedInfo.

Move CryptPrivateKey and MakeFake*Key functions
into kbfscrypto.

Port some RootMetadata tests to BareRootMetadata
tests.

Do device name lookups at conflict renaming time instead
of at construction of writerInfo time.

Fix some constant types.

Lots of test cleanup.

### CODE REVIEW

**Comment by akalin-keybase** [comment]
@aalness please review changes to BareRootMetadata and RootMetadata (also may want to look at https://github.com/keybase/kbfs/pull/474/files concurrently). @strib please review everything else.

**Comment by aalness** [comment]
Looks great to me.

File: libkbfs/conflict_resolver.go [hunk 1]
```diff
@@ -374,7 +374,7 @@ func fileWithConflictingWrite(unmergedChains, mergedChains *crChains,
 // unmerged paths that need to be checked for conflicts later in
 // conflict resolution, for all subdirectories of the given path.
 func (cr *ConflictResolver) checkPathForMerge(ctx context.Context,
-	unmergedChain *crChain, unmergedPath path, unmergedChains *crChains,
+	unmergedChain *crChain, unmergedPath path, unmergedChains,
```
Comment (strib) [inline libkbfs/conflict_resolver.go]: I find it really confusing when the type of a param is split into the next line.  Can you please keep this as is, or moved `unmergedChains` down to the next line? (And all the places below.)

File: libkbfs/folder_branch_ops.go [hunk 2]
```diff
@@ -92,8 +92,8 @@ type fboMutexLevel mutexLevel
 
 const (
 	fboMDWriter fboMutexLevel = 1
-	fboHead                   = 2
-	fboBlock                  = 3
+	fboHead     fboMutexLevel = 2
```
Comment (strib) [inline libkbfs/folder_branch_ops.go]: Why?  I thought the previous way was idiomatic.  (Same with the `MetadataVer` stuff above.)

File: libkbfs/kbfs_ops_test.go [hunk 3]
```diff
@@ -278,13 +278,16 @@ func getOps(config Config, id TlfID) *folderBranchOps {
 		getOpsNoAdd(FolderBranch{id, MasterBranch})
 }
 
+// TODO: Test MDv3.
```
Comment (strib) [inline libkbfs/kbfs_ops_test.go]: Won't this happen automatically when we flip the default version in the config?

File: libkbfs/conflict_renamer.go [hunk 4]
```diff
@@ -60,18 +67,11 @@ func splitExtension(path string) (string, string) {
...
-func newWriterInfo(ctx context.Context, cfg Config, uid keybase1.UID,
-	key kbfscrypto.VerifyingKey,
-	revision MetadataRevision) (writerInfo, error) {
-	ui, err := cfg.KeybaseService().LoadUserPlusKeys(ctx, uid)
-	if err != nil {
-		return writerInfo{}, err
-	}
-
+func newWriterInfo(uid keybase1.UID, key kbfscrypto.VerifyingKey,
+	revision MetadataRevision) writerInfo {
```
Comment (akalin-keybase) [inline libkbfs/conflict_renamer.go]: For posterity, some justification for this, as I have vague memories of pushing for the old version (do all the device name loading at `newWriterInfo` time). That was okay when conflict resolution was the only thing using `crChains`, but it seems like `crChains` is a more general-purpose data structure, and those other use cases don't need to do any resolving, so it makes more sense to do the device name loading on-demand.

File: libkbfs/folder_branch_ops.go (hunk 2 above)
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: (Answered in chat, but also answering here for posterity.) Unfortunately, it's not. Without the explicit types, the subsequent constants are untyped which basically means they behave like ints.

The relevant part of the spec is from https://golang.org/ref/spec#Constant_declarations :

```
If the type is omitted, the constants take the individual types of the corresponding expressions. If the expression values are untyped constants, the declared constants remain untyped and the constant identifiers denote the constant values.
...
Within a parenthesized const declaration list the expression list may be omitted from any but the first declaration. Such an empty list is equivalent to the textual substitution of the first preceding non-empty expression list and its type if any. Omitting the list of expressions is therefore equivalent to repeating the previous list. The number of identifiers must be equal to the number of expressions in the previous list. Together with the iota constant generator this mechanism permits light-weight declaration of sequential values...
```

Which explains why they're untyped in general, but are actually typed in, e.g.:

```
const (
  Foo MyType = iota
  Bar
  Baz
)
```

File: libkbfs/conflict_resolver.go (hunk 1 above)
Comment (akalin-keybase) [inline libkbfs/conflict_resolver.go]: Done, moved to next line (and everywhere else). Add more lines, but I think it's easier to read anyway.

**Pull request merged**